lib/puppet/transaction/event.rb

```diff
@@ -94,6 +98,30 @@ def to_yaml_properties
     YAML_ATTRIBUTES & super
   end
 
+  # Calculate and set the corrective_change parameter, based on the old_system_value of the property.
+  # @param [Object] old_system_value system_value from last transaction
+  # @return [bool] true if this is a corrective_change
+  def calculate_corrective_change(old_system_value)
+    # Only idempotent properties, and cases where we have an old system_value
+    # are corrective_changes.
+    if @property_instance.idempotent? &&
+       !@property_instance.sensitive &&
+       !old_system_value.nil?
+
+      # Here we duplicate the property to avoid side-effects, and modify its should value
+      # so we can use insync?
+      test_property = @property_instance.dup
```

Comment:
This scares the crap out of me. The RAL isn't intended to duplicate properties like this, and while this is possible it's working against how things are built. My biggest concern is that I can't predict how this will behave but I do suspect that this might not be safe.

/cc @branan